curvefs/test/metaserver/copyset/concurrent_apply_queue_test.cpp

```diff
@@ -0,0 +1,222 @@
+/*
+ *  Copyright (c) 2023 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/*
+ * Project: curve
+ * File Created: 20230521
+ * Author: Xinlong-Chen
+ */
+
+#include <gtest/gtest.h>
+
+#include <vector>
+#include <atomic>
+#include <functional>
+
+#include "src/common/timeutility.h"
+#include "curvefs/src/metaserver/copyset/concurrent_apply_queue.h"
+
+using curvefs::metaserver::copyset::ApplyQueue;
+using curvefs::metaserver::copyset::ApplyOption;
+using curvefs::metaserver::copyset::OperatorType;
+
+TEST(ApplyQueue, InitTest) {
+    ApplyQueue concurrentapply;
+
+    {
+        // 1. init with invalid write-concurrentsize
+        ApplyOption opt(-1, 1, 1, 1);
+        ASSERT_FALSE(concurrentapply.Init(opt));
+    }
+
+    {
+        // 2. init with invalid write-concurrentdepth
+        ApplyOption opt(1, -1, 1, 1);
+        ASSERT_FALSE(concurrentapply.Init(opt));
+    }
+
+    {
+        // 3. init with invalid read-concurrentsize
+        ApplyOption opt(1, 1, -1, 1);
+        ASSERT_FALSE(concurrentapply.Init(opt));
+    }
+
+    {
+        // 4. init with invalid read-concurrentdepth
+        ApplyOption opt(1, 1, 1, -1);
+        ASSERT_FALSE(concurrentapply.Init(opt));
+    }
+
+    {
+        // 5. double init
+        ApplyOption opt(1, 1, 1, 1);
+        // init with vaild params
+        ASSERT_TRUE(concurrentapply.Init(opt));
+        // re-init
+        ASSERT_TRUE(concurrentapply.Init(opt));
+    }
+
+    concurrentapply.Stop();
+    // re-stop
+    concurrentapply.Stop();
+}
+
+void init_read_write_type_list(std::vector<OperatorType> *read_type_list,
```

Comment:
Please follow Google C++ code style, for class and function names, they should be CameCase, and for variables, they should be camelCase.

And consider putting functions that only used in current TU under unnamed namespace or mark them static.

Reply:
done